cranelift-codegen/meta/src/isa/x86/encodings.rs

```diff
@@ -177,16 +187,18 @@ impl PerCpuModeEncodings {
     }
 
     /// Add encodings for `inst.r32` to X86_32.
-    /// Add encodings for `inst.r32` to X86_64 with and without REX.
     /// Add encodings for `inst.r64` to X86_64 with a REX.W prefix.
     fn enc_r32_r64(&mut self, inst: impl Into<InstSpec>, template: Template) {
         let inst: InstSpec = inst.into();
         self.enc32(inst.bind_ref(R32), template.nonrex());
+        self.enc64(inst.bind_ref(R64), template.rex().w());
+    }
 
-        // REX-less encoding must come after REX encoding so we don't use it by default. Otherwise
-        // reg-alloc would never use r8 and up.
-        self.enc64(inst.bind_ref(R32), template.rex());
-        self.enc64(inst.bind_ref(R32), template.nonrex());
+    /// Add encodings for `inst.r32` to X86_32.
+    /// Add encodings for `inst.r64` to X86_64 with a REX.W prefix.
+    fn enc_r32_r64_rex_only(&mut self, inst: impl Into<InstSpec>, template: Template) {
+        let inst: InstSpec = inst.into();
+        self.enc32(inst.bind_ref(R32), template.nonrex());
         self.enc64(inst.bind_ref(R64), template.rex().w());
     }
```

Comment:
No difference with `enc_r32_r64`?

Reply:
Ah, well spotted.  I removed `enc_r32_r64` and replaced all uses of it with `enc_r32_r64_rex_only`.  The alternative would have been to remove `enc_r32_r64_rex_only` and use `enc_r32_r64` instead, but that would mean that the REX behaviour of `enc_r32_r64` was inconsistent with that of the otherwise very similar `enc_i32_i64`.  So I didn't do that.